pam/pam_test/module-transaction-dummy_test.go

```diff
@@ -0,0 +1,512 @@
+				err := tx.SetItem(pam.User, tc.presetUser)
+				require.NoError(t, err)
+			}
+
+			prompt := "Who are you?"
+			user, err := tx.GetUser(prompt)
+			require.ErrorIs(t, err, tc.expectedError)
+			require.Equal(t, tc.expected, user)
+		})
+	}
+}
+
+func TestStartStringConv(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		prompt                string
+		promptFormat          string
+		promptFormatArgs      []interface{}
+		convStyle             pam.Style
+		convError             error
+		convHandler           *pam.ConversationFunc
+		convShouldNotBeCalled bool
+		expected              string
+		expectedError         error
+	}{
+		"normal string": {
+			prompt:    "This is an error!",
+			convStyle: pam.ErrorMsg,
+			expected:  "I'm handling it fine though",
+		},
+
+		"formatted string": {
+			promptFormat:     "Sending some %s, right? %v",
+			promptFormatArgs: []interface{}{"info", true},
+			convStyle:        pam.TextInfo,
+			expected:         "And returning some text back",
+		},
+
+		"failing conv func": {
+			prompt:        "Tell me your secret!",
+			convStyle:     pam.PromptEchoOff,
+			convError:     pam.BufErr,
+			expectedError: pam.BufErr,
+		},
+
+		"no conv set fails": {
+			convHandler:   ptrValue(pam.ConversationFunc(nil)),
+			expectedError: pam.ConvErr,
+		},
+
+		"failing conv func per invalid style": {
+			prompt:                "Tell me your secret!",
+			convStyle:             pam.BinaryPrompt,
+			convError:             pam.ConvErr,
+			expectedError:         pam.ConvErr,
+			convShouldNotBeCalled: true,
+		},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(MaybeDoLeakCheck)
+
+			convFunCalled := false
+			tx := NewModuleTransactionDummy(func() pam.ConversationFunc {
+				if tc.convHandler != nil {
+					return *tc.convHandler
+				}
+				prompt := tc.prompt
+				if tc.promptFormat != "" {
+					prompt = fmt.Sprintf(tc.promptFormat, tc.promptFormatArgs...)
+				}
+				return pam.ConversationFunc(
+					func(style pam.Style, msg string) (string, error) {
+						convFunCalled = true
+						require.Equal(t, prompt, msg)
+						require.Equal(t, tc.convStyle, style)
+						return tc.expected, tc.convError
+					})
+			}())
+
+			var reply pam.StringConvResponse
+			var err error
+
+			if tc.promptFormat != "" {
+				reply, err = tx.StartStringConvf(tc.convStyle, tc.promptFormat,
+					tc.promptFormatArgs...)
+			} else {
+				reply, err = tx.StartStringConv(tc.convStyle, tc.prompt)
+			}
+
+			require.ErrorIs(t, err, tc.expectedError)
+			expectConFuncCalled := !tc.convShouldNotBeCalled && tc.convHandler == nil
+			require.Equal(t, expectConFuncCalled, convFunCalled)
+
+			if tc.expectedError == nil {
+				require.NotNil(t, reply)
+				require.Equal(t, tc.expected, reply.Response())
+				require.Equal(t, tc.convStyle, reply.Style())
+			} else {
+				require.Zero(t, reply)
+			}
```

Comment:
It's better to keep the final assertions as together as possible and we can avoid the `else` as well:
```suggestion
			expectConFuncCalled := !tc.convShouldNotBeCalled && tc.convHandler == nil
			
			require.Equal(t, expectConFuncCalled, convFunCalled)
			require.ErrorIs(t, err, tc.expectedError)
			if tc.expectedError != nil {
				require.Zero(t, reply)
				return
			}
			require.NotNil(t, reply)
			require.Equal(t, tc.expected, reply.Response())
			require.Equal(t, tc.convStyle, reply.Style())
```

Reply:
Mh, I was thinking about returning early here... But while I like the way I feel it to be a bit dangerous at times, given that if you want to add more checks, then you may forget about doing further tests, but that's fine.